File: 19/cpp/main.cpp
#include <iostream>
#include <stack>

class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        ListNode* return_head = head;
        std::stack<ListNode*> stack;

        while (head != nullptr) {
            stack.push(&head);
            head = head->next;
        }

        for (int i = 1; i <= n; i++) {
            stack.pop();
        }

        // If the stack is empty, it means we are removing the head node
        if (stack.empty()) {
            // Return the next node as the new head
            return return_head->next;
        }

        ListNode* end_normal = stack.top();
        end_normal->next = end_normal->next->next;

        return return_head;
    }
};

int main() {
    Solution sol;

    // Test case X: Y
    RETURN_TYPE resultX = sol.FUNCTION_NAME(Y);
    std::cout << "Test X: " << resultX << std::endl;

    return 0;
}

File: 19/python3/main.py
#!/usr/bin/env python3

class Solution:
    def removeNthFromEnd(self, head: ListNode, n: int) -> ListNode:
        # Use a stack to store pointers to the nodes
        stack = []
        current = head

        # Push each node onto the stack
        while current:
            stack.append(current)
            current = current.next

        # Pop n nodes from the stack to find the nth node from the end
        for _ in range(n):
            stack.pop()

        # If the stack is empty, we're removing the head node
        if not stack:
            return head.next

        # Get the node before the nth node from the end
        node_before_nth = stack[-1]

        # Remove the nth node from the end
        node_before_nth.next = node_before_nth.next.next

        return head
